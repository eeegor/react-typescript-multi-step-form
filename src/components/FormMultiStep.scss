@import '../variables';

.form__title {
	margin-top: 0;
	margin-bottom: 1rem;
	font-size: 2rem;
}

.form-multi-step__next {
	margin: 2rem 0;
}

.form-multi-step__success {
	text-align: center;
}
