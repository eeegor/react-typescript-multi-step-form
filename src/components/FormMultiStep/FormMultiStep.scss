.form-multi-step {
	margin-bottom: 2rem;
}

.form-multi-step__next {
	margin: 2rem auto;
}
