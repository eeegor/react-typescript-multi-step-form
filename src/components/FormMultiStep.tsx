/**
 * Component FormMultiStep
 */

import * as React from 'react';
import { FormStep } from './FormStep';
import { InputGroupText } from './InputGroupText';
import { InputSelectRadio } from './InputSelectRadio';
import { Props as InputGroupTextProps } from './InputGroupText';
import { Props as InputSelectRadioProps } from './InputSelectRadio';
import './FormMultiStep.scss';

export interface FormSchema {
	name: string;
	email: string;
	phone: string;
	salary: string;
}

export interface Props {
	id: string;
	// tslint:disable-next-line:no-any
	formSchema: { [key: number]: any };
	onChange: (form: FormSchema) => void;
}

interface State {
	showDebugState: boolean;
	currentStep: number;
	submit: boolean;
	form: FormSchema;
}

const DebugState = (props: { state: State; showDebugState: boolean; onClick: () => void }) => (
	<div className="debug-state">
		<div className="debug-state__toggle" onClick={() => props.onClick()}>
			Show state
		</div>
		{props.showDebugState && (
			<div className="debug-state__body">
				<h3>State</h3>
				{Object.keys(props.state).map(stateItem => (
					<div className="flex" key={stateItem}>
						<label>{stateItem}</label>
						<div>
							<pre>
								<code>{JSON.stringify(props.state[stateItem], null, 4)}</code>
							</pre>
						</div>
					</div>
				))}
			</div>
		)}
	</div>
);

export class FormMultiStep extends React.Component<Props, State> {
	state = {
		showDebugState: false,
		currentStep: 1,
		submit: false,
		form: {
			name: '',
			email: '',
			phone: '',
			salary: '',
		},
	};

	handleFormSubmit = (form: React.FormEvent<HTMLFormElement>) => {
		form.preventDefault();
		this.props.onChange(this.state.form);
	};

	handleInputChange = (name: string, value: string) => {
		this.setState(state => ({
			...state,
			form: {
				...state.form,
				[name]: value,
			},
		}));
	};

	toggleDebugState = () =>
		this.setState(state => ({ ...state, showDebugState: !state.showDebugState }));

	render(): JSX.Element {
		const { id, formSchema } = this.props;
		const { form, currentStep, showDebugState } = this.state;

		return (
			<div id={`form-multi-step--${id}`} className="form-multi-step">
				<div className="container">
					<h1>Multi Step Form</h1>
					<p>
						Current step:{' '}
						<span>
							{currentStep} / {Object.keys(form).length}
						</span>
					</p>

					<DebugState
						state={this.state}
						showDebugState={showDebugState}
						onClick={() => this.toggleDebugState()}
					/>

					{Object.keys(formSchema).map(formSchemaStep => {
						const formStepData = formSchema[formSchemaStep];
						return (
							<FormStep
								key={formSchemaStep}
								id={`form-step-${formSchemaStep}`}
								onSubmit={formData => this.handleFormSubmit(formData)}
							>
								{formStepData.type === 'select-radio' && (
									<InputSelectRadio
										id={formStepData.id}
										name={formStepData.name}
										label={formStepData.label}
										values={formStepData.values}
										defaultChecked={
											(formStepData.values.includes(
												form[formStepData.name]
											) &&
												form[formStepData.name]) ||
											formStepData.values[0]
										}
										onChange={(name, event) =>
											this.handleInputChange(name, event.target.value)
										}
									/>
								)}
								{(formStepData.type === 'text' ||
									formStepData.type === 'email') && (
									<InputGroupText
										name={formStepData.name}
										label={formStepData.label}
										type={formStepData.type}
										value={form[formStepData.name]}
										onChange={(name, event) =>
											this.handleInputChange(name, event.target.value)
										}
									/>
								)}
							</FormStep>
						);
					})}
				</div>
			</div>
		);
	}
}
