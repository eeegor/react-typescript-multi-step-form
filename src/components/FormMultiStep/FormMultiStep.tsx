/**
 * Component FormMultiStep
 */

import React from 'react';
import classnames from 'classnames';
import {
	FormStep,
	InputControlText,
	InputControlSelectRadio,
	InputControlTextProps,
	InputControlSelectRadioProps,
	InputControlSubmitProps,
	InputControlSubmit,
	FormControls,
	FormSuccess,
	Button,
} from '..';
import './FormMultiStep.scss';

export type FormMultiStepAllowedTypes =
	| InputControlTextProps
	| InputControlSelectRadioProps
	| InputControlSubmitProps;

export type FormMultiStepSchema = {
	[key: number]: FormMultiStepAllowedTypes;
};

export interface Props {
	id: string;
	className?: string;
	formSchema: object;
	formData?: object;
	onChange: (state: State) => void;
}

interface State {
	currentStep: number;
	complete: boolean;
	valid: boolean;
	form: object;
	errors: object;
}

const makeDefaultFormData = (formSchema: object): object => {
	let nextFormData = {};
	Object.values(formSchema).map(field => {
		switch (field.type) {
			case 'submit':
				return (nextFormData[field.name] = false);
			case 'radio':
			case 'text':
			case 'email':
			default:
				return (nextFormData[field.name] = '');
		}
	});
	return nextFormData;
};

const prepareFormData = (formData?: object): object | undefined => {
	if (formData && !formData.hasOwnProperty('submit')) {
		return { ...formData, submit: false };
	}
	return formData || undefined;
};

const initialState = (formSchema: object, formData?: object): State => {
	return {
		currentStep: 1,
		complete: false,
		valid: false,
		form: prepareFormData(formData) || makeDefaultFormData(formSchema) || {},
		errors: {},
	};
};

const validateString = (value: string): boolean => /[0-9a-zA-Z]{2,}/i.test(`${value}`);
const validateEmail = (value: string): boolean =>
	/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(
		`${value}`
	);

const getValidationMessage = (current: object, state: State): string => {
	switch (current['type']) {
		case 'radio':
			return 'You need to select one item';
		case 'email':
			return 'Email is required';
		case 'text':
		default:
			return 'This field is required';
	}
};

export class FormMultiStep extends React.Component<Props, State> {
	state: State = initialState(this.props.formSchema, this.props.formData);

	gotoPrevStep = (): void => {
		this.setState(state => ({
			...state,
			complete: false,
			currentStep: state.currentStep > 1 ? state.currentStep - 1 : 1,
		}));
	};

	gotoNextStep = (): void => {
		const maxSteps = Object.keys(this.state.form).length;

		if (this.validateInput()) {
			this.setState(state => ({
				...state,
				complete: false,
				currentStep: state.currentStep < maxSteps ? state.currentStep + 1 : maxSteps,
			}));
		}
	};

	gotoStep = (step: number): void => {
		const maxSteps = Object.keys(this.state.form).length;
		this.setState(state => ({
			...state,
			complete: false,
			currentStep: step > 0 ? (step < maxSteps ? step : maxSteps) : 1,
		}));
	};

	validateInput = (): boolean => {
		const current = this.props.formSchema[this.state.currentStep];
		const hasErrors = this.state.form[current.name] === '';
		const invalidString =
			current['type'] === 'text' && !validateString(this.state.form[current.name]);
		const invalidEmail =
			current['type'] === 'email' && !validateEmail(this.state.form[current.name]);

		if (hasErrors || invalidEmail || invalidString) {
			this.setState(state => ({
				...state,
				errors: {
					...this.state.errors,
					[current.name]:
						(hasErrors && getValidationMessage(current, this.state)) ||
						(invalidEmail && 'Email is not valid') ||
						(invalidString && 'This entry seems not valid...') ||
						'Something went wrong',
				},
			}));
			return false;
		}
		return true;
	};

	handleFormSubmit = (form: React.FormEvent<HTMLFormElement>): void => {
		form.preventDefault();
		this.gotoNextStep();
		this.props.onChange(this.state);
	};

	handleInputChange = (name: string, value: string): void => {
		this.setState(
			state => ({
				...state,
				errors: {},
				valid: false,
				form: {
					...state.form,
					[name]: value,
				},
			}),
			() => this.props.onChange(this.state)
		);
	};

	handleFormComplete = (): void => {
		this.setState(state => ({
			...state,
			complete: true,
		}));
		this.props.onChange(this.state);
	};

	resetForm = (event: React.MouseEvent<HTMLButtonElement>) => {
		event.preventDefault();
		this.setState(initialState(this.props.formSchema, this.props.formData), () =>
			this.props.onChange(this.state)
		);
	};

	render(): JSX.Element {
		const { id, className, formSchema } = this.props;
		const { form, currentStep, complete, valid } = this.state;
		const input = formSchema[currentStep];
		const maxSteps = Object.keys(form).length;
		const hasErrors = this.state.errors[input.name];

		return (
			<div id={`form-multi-step-${id}`} className={classnames('form-multi-step', className)}>
				{complete && <FormSuccess onReset={event => this.resetForm(event)} />}
				{!complete && (
					<>
						<FormControls
							maxSteps={maxSteps}
							currentStep={currentStep}
							onBack={() => this.gotoPrevStep()}
						/>
						<FormStep
							key={currentStep}
							id={`form-step-${currentStep}`}
							onSubmit={formData => this.handleFormSubmit(formData)}
						>
							{input.type === 'radio' && (
								<InputControlSelectRadio
									id={input.id}
									required={input.required}
									name={input.name}
									label={input.label}
									values={input.values}
									info={input.info}
									status={input.status}
									errors={hasErrors}
									defaultChecked={
										(input.values.includes(form[input.name]) &&
											form[input.name]) ||
										(input.defaultChecked !== false && input.values[0])
									}
									onChange={(name, event) =>
										this.handleInputChange(name, event.target.value)
									}
								/>
							)}
							{(input.type === 'text' || input.type === 'email') && (
								<InputControlText
									id={input.id}
									required={input.required}
									name={input.name}
									label={input.label}
									type={input.type}
									value={form[input.name]}
									info={input.info}
									errors={hasErrors}
									valid={valid}
									status={input.status}
									onChange={(name, event) =>
										this.handleInputChange(name, event.target.value)
									}
								/>
							)}
							{input.type === 'submit' && (
								<InputControlSubmit
									id={input.id}
									required={input.required}
									name={input.name}
									label={input.label}
									value={input.value}
									info={input.info}
									status={input.status}
									formValues={form}
									errors={hasErrors}
									onClick={() => this.handleFormComplete()}
									gotoStep={step => this.gotoStep(step)}
								/>
							)}
							{input.type !== 'submit' && (
								<Button
									size="large"
									className="form-multi-step__next"
									type={(currentStep === maxSteps && 'disabled') || 'secondary'}
									onClick={() => this.gotoNextStep()}
								>
									Next
								</Button>
							)}
						</FormStep>
					</>
				)}
			</div>
		);
	}
}
