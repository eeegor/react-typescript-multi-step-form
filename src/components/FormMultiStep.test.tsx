import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { FormMultiStep } from './FormMultiStep';

describe('FormMultiStep', () => {
	const callback = jest.fn();
	const testData = {
		1: { type: 'text', id: 'name', name: 'name', label: 'Your full name', value: 'Demo' },
		2: {
			type: 'email',
			id: 'email',
			name: 'email',
			label: 'Your email',
			value: 'info@example.com',
		},
		3: {
			type: 'text',
			id: 'phone',
			name: 'phone',
			label: 'Your phone',
			value: '+49 160 3232323',
		},
		4: {
			type: 'select-radio',
			id: 'salary',
			name: 'salary',
			label: 'Select you salary',
			values: [
				'€ 0 - € 1.000',
				'€ 1.000 - € 2.000',
				'€ 2.000 - € 3.000',
				'€ 3.000 - € 4.000',
				'More then € 4.000',
			],
		},
	};

	it('renders markup correctly', () => {
		const component = <FormMultiStep formSchema={testData} onChange={callback} />;
		const tree = renderer.create(component).toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe('Form Text Fields', () => {
		const testDataForm = [
			{ type: 'text', name: 'name', value: 'Demo' },
			{ type: 'text', name: 'email', value: 'info@example.com' },
			{ type: 'text', name: 'phone', value: '+49 160 3232323' },
			{
				type: 'select-radio',
				name: 'salary',
				value: '1.000 - 2.000',
			},
		];

		testDataForm.forEach((testSet, index) => {
			it(`can change and submit ${testSet.name} input`, () => {
				const nextValue = { target: { value: testSet.value } };
				const wrapper = mount(<FormMultiStep formSchema={testData} onChange={callback} />);
				const form = wrapper.find(`#form-step-${index + 1}`);
				switch (testSet.type) {
					case 'select-radio':
						const formRadioGroup = form.find('.input-select--radio');
						const formSelectRadioInput = formRadioGroup.find('.input-radio').first();
						formSelectRadioInput.simulate('focus');
						formSelectRadioInput.simulate('change', nextValue);
						formRadioGroup.simulate('submit', nextValue);
						break;

					case 'text':
					default:
						const formTextGroup = form.find('.input-text');
						formTextGroup.simulate('focus');
						formTextGroup.simulate('change', nextValue);
						formTextGroup.simulate('submit', nextValue);
						break;
				}
				const formState = wrapper.state()['form'];
				expect(formState[testSet.name]).toEqual(testSet.value);
			});
		});
	});
});
