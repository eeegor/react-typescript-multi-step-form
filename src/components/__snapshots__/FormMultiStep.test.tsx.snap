// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormMultiStep renders markup correctly 1`] = `
<div
  className="form-multi-step"
  id="form-multi-step--test"
>
  <div
    className="container"
  >
    <h1>
      Multi Step Form
    </h1>
    <div>
      <div
        className="flex"
      >
        <label>
          name
        </label>
        <div>
          No data yet...
        </div>
      </div>
      <div
        className="flex"
      >
        <label>
          email
        </label>
        <div>
          No data yet...
        </div>
      </div>
      <div
        className="flex"
      >
        <label>
          phone
        </label>
        <div>
          No data yet...
        </div>
      </div>
      <div
        className="flex"
      >
        <label>
          salary
        </label>
        <div>
          No data yet...
        </div>
      </div>
    </div>
    <hr />
    <form
      className="form form-step"
      id="form-step-1"
      onSubmit={[Function]}
    >
      <div
        className="form__container"
      >
        <div
          className="input-group input-group--text"
        >
          <label
            className="input-group__label"
            htmlFor="name"
          >
            Your full name
          </label>
          <input
            className="input-group__input input-text"
            id="name"
            name="name"
            onChange={[Function]}
            type="text"
            value="Demo"
          />
        </div>
      </div>
    </form>
    <form
      className="form form-step"
      id="form-step-2"
      onSubmit={[Function]}
    >
      <div
        className="form__container"
      >
        <div
          className="input-group input-group--email"
        >
          <label
            className="input-group__label"
            htmlFor="email"
          >
            Your email
          </label>
          <input
            className="input-group__input input-text"
            id="email"
            name="email"
            onChange={[Function]}
            type="email"
            value="info@example.com"
          />
        </div>
      </div>
    </form>
    <form
      className="form form-step"
      id="form-step-3"
      onSubmit={[Function]}
    >
      <div
        className="form__container"
      >
        <div
          className="input-select input-select--radio"
        >
          <label
            className="label"
            htmlFor="salary"
          >
            Select you salary
          </label>
          <div
            className="input-group input-group--radio"
          >
            <input
              className="input-group__input input-radio"
              defaultChecked={true}
              id="salary-0"
              name="salary"
              onChange={[Function]}
              type="radio"
              value="€ 0 - € 1.000"
            />
            <label
              className="input-group__label"
              htmlFor="salary-0"
            >
              € 0 - € 1.000
            </label>
          </div>
          <div
            className="input-group input-group--radio"
          >
            <input
              className="input-group__input input-radio"
              defaultChecked={false}
              id="salary-1"
              name="salary"
              onChange={[Function]}
              type="radio"
              value="€ 1.000 - € 2.000"
            />
            <label
              className="input-group__label"
              htmlFor="salary-1"
            >
              € 1.000 - € 2.000
            </label>
          </div>
          <div
            className="input-group input-group--radio"
          >
            <input
              className="input-group__input input-radio"
              defaultChecked={false}
              id="salary-2"
              name="salary"
              onChange={[Function]}
              type="radio"
              value="€ 2.000 - € 3.000"
            />
            <label
              className="input-group__label"
              htmlFor="salary-2"
            >
              € 2.000 - € 3.000
            </label>
          </div>
          <div
            className="input-group input-group--radio"
          >
            <input
              className="input-group__input input-radio"
              defaultChecked={false}
              id="salary-3"
              name="salary"
              onChange={[Function]}
              type="radio"
              value="€ 3.000 - € 4.000"
            />
            <label
              className="input-group__label"
              htmlFor="salary-3"
            >
              € 3.000 - € 4.000
            </label>
          </div>
          <div
            className="input-group input-group--radio"
          >
            <input
              className="input-group__input input-radio"
              defaultChecked={false}
              id="salary-4"
              name="salary"
              onChange={[Function]}
              type="radio"
              value="More then € 4.000"
            />
            <label
              className="input-group__label"
              htmlFor="salary-4"
            >
              More then € 4.000
            </label>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
